src/platform_linux/platform.h

```diff
@@ -747,51 +776,199 @@ platform_enable_tracing_shm_frees();
+ *   freed).
+ *
+ * - Defenses are built-in to protect callers which may be incorrectly using
+ *   this interface to free memory allocated from shared-segment or the heap.
+ * ----------------------------------------------------------------------------
  */
-#define platform_free(id, p)                                                   \
+#define platform_free(hid, p)                                                  \
+   do {                                                                        \
+      platform_assert(((p) != NULL),                                           \
+                      "Attempt to free a NULL ptr from '%s', line=%d",         \
+                      __func__,                                                \
+                      __LINE__);                                               \
+      if (IS_MEM_FRAG(p)) {                                                    \
+         platform_memfrag *_mf = (platform_memfrag *)(p);                      \
+         platform_free_mem((hid), _mf->addr, _mf->size, STRINGIFY(p));         \
+         _mf->addr = NULL;                                                     \
+         _mf->size = 0;                                                        \
+      } else {                                                                 \
+         platform_assert((((hid) == PROCESS_PRIVATE_HEAP_ID)                   \
+                          || (sizeof(*p) > sizeof(uint64))),                   \
+                         "Attempt to free memory using '%s', which is a "      \
+                         " pointer to an object of size %lu bytes"             \
+                         ", from '%s', line=%d\n",                             \
+                         STRINGIFY(p),                                         \
+                         sizeof(*p),                                           \
+                         __func__,                                             \
+                         __LINE__);                                            \
+         /*                                                                    \
+          * Expect that 'p' is pointing to a struct. So get its size.          \
+          * Except that while allocating memory for objects, we had previously \
+          * aligned the required size to PLATFORM_CACHELINE_SIZE. Redo that    \
+          * work here, so we correctly tell shared-memory the size to free.    \
+          */                                                                   \
+         const size_t _size = sizeof(*p);                                      \
+         const size_t _reqd =                                                  \
+            (_size + platform_alignment(PLATFORM_CACHELINE_SIZE, _size));      \
+         platform_free_mem((hid), (p), _reqd, STRINGIFY(p));                   \
+         (p) = NULL;                                                           \
```

Comment:
@rtjohnso - This line and L907 below is what makes it necessary for clients calling `free()` to do two things:

```
   platform_memfrag memfrag;
   platform_memfrag *mf;
   ... Do the initialization ...
  platform_free(mf);
```

I would have liked to skip the `mf` and simply pass-in `&memfrag` to `free()`, but there is a compiler error.

I think this can be fixed with some rework ... but I ran out of energy. Let's review if this can be improved.